README
     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================


                	HadrHN test     

                  V. Ivanchenko, A. Bagulya, B. Kutsenko

The goal of this test is to identify accuracy of Geant4 elementary cross sections 
versus PDG collection of data and to control the hydrogen cross section from cross 
sections classes. These elementary cross sections are used to compute cross section 
off nuclei and are also used in hadronic models.

Projectile: proton, neutron, pi+, pi-, kaon+, kaon-
Elementary cross sections: “NS”, “PDG16”, “SAID”
Cross sections classes: “CHIPS”, “BGG”, “XS”
 
Data are from  http://pdg.lbl.gov.
Files with data are located in HadrHN/data directory. 

	        Compilation
G4BLOCAL points to directory where geant4-dev is located
HADSRC -> .../g4tests-verification/hadronic
G4BWORK points to directory where G4 executables are located (.../bin)
G4BBUILD points to build directory for geant4 (.../geant4-build-mt/)
G4VERSION ->  (10.5.0)

mkdir -p ${G4BLOCAL}/tests-build/HadrHN
cd ${G4BLOCAL}/tests-build/HadrHN
rm -f HadrHN

cmake -DCMAKE_INSTALL_PREFIX=${G4BLOCAL} \
      -DGeant4_DIR=${G4BBUILD}/lib64/Geant4-${G4VERSION} ${HADSRC}/HadrHN
make -j2

cd ${G4BWORK}
rm -f HadrHN
ln -s ${G4BLOCAL}/tests-build/HadrHN/HadrHN

cd ${HADSRC}/HadrHN

                Run Productuion

Simple run for proton, neutron:
$G4BWORK/HadrHN p NS [PDG16/SAID/BGG/CHIPS/XS]
Simple run for pi+, pi-:
$G4BWORK/HadrHN pi+ NS [PDG16/SAID/BGG/CHIPS] 
Simple run for kaon:
$G4BWORK/HadrHN kaon+ kNS [VG/GG/CHIPS]
  here VG - work only for K+-
       GG - for all kaons using VG
$G4BWORK/HadrHN alpha XS [ION]

               Histogramms
ROOT files are creating automatically in HadrHN/test directory.

		Plots
python plotXS.py proton_proton_NS.root 0 # each ROOT file from HadrHN/test directory
python plotXS.py proton_proton_NS.root 1 # "1" only for p+p interaction to add cosmic data

Plots are creating in HadrHN/plotXS directory.



History
$Id: History,v 1.92 2010-11-09 21:26:24 asaim Exp $
----------------------------------------------------

     =========================================================
     Geant4 - an Object-Oriented Toolkit for Simulation in HEP
     =========================================================

                    HadrHN History file
                    ---------------------
This file should be used by the G4 example coordinator to briefly
summarize all major modifications introduced in the code and keep
track of all tags.

     ----------------------------------------------------------
     * Reverse chronological order (last date on top), please *
     ----------------------------------------------------------

18-04-19 A.Bagulia (hadrhn-V10-05-02)
- added README

09-04-19 A.Bagulia (hadrhn-V10-05-01)
- updated HadrHN.cc, plotXS.py

29-11-18 V.Ivanchenko (hadrhn-V10-04-01)
- for Geant4 10.5

22-05-18 V.Ivanchenko (hadrhn-V10-04-00)
     


plotXS.C
// pictures XS
// 
// 06.07.2012


string sInt(Int_t number)
  {
    stringstream ss;
    ss << number;
    return ss.str();
  }

void plotXS()
{
  cout <<  "Started plotXS() " << endl;
  int verbose = 0;
  gROOT->SetStyle("Plain");
  gROOT->Reset();
  static const G4int NVAR=16;
  static const G4int NVAR1=48;
  G4String type[NVAR] = {"Total p + p","Total n + p","Total pi+ + p",
			 "Total pi- + p","Total p + n","Total n + n",
			 "Total pi+ + n","Total pi- + n",
			 "Elastic p + p","Elastic n + p","Elastic pi+ + p",
			 "Elastic pi- + p","Elastic p + n","Elastic n + n",
			 "Elastic pi+ + n","Elastic pi- + n"};
  G4String num[NVAR1] = {"0","1","2","3","4","5","6","7",
			"8","9","10","11","12","13","14","15",
			"16","17","18","19","20","21","22","23",
			"24","25","26","27","28","29","30","31",
			"32","33","34","35","36","37","38","39",
			"40","41","42","43","44","45","46","47"};


  int ee[5] = {0, 1, 10, 100, 250};
  double Xmin[5] = {0,-1,-1,-1,-1};
  double Xmax[5] = {8,1,1,1,1};
  double Ymin[5] = {0.00015, 0., 1.1, 1.1, 1.1};
  double Ymax[5] = {.015,1000,10000,10000,10000};
  double Y1max[101]={0,
  .1,.1,.15,.15,.5,.5,.5,.5,.5,.5,  // 1-10
  .15,.15,.15,.15,.15,.15,.15,.15,.15,.15,  // 11-20
  .15,.15,.15,.15,.15,.15,.15,.15,.15,.15,  // 21-30
  .15,.15,.15,.15,.15,.15,.15,.15,.15,.15,  // 31-40
  .15,.15,.15,.15,.15,.15,.15,.15,.15,.15,  // 41-50
  .15,.15,.15,.15,.15,.15,.15,.15,.15,.15,  // 51-60
  .15,.15,.15,.15,.15,.15,.15,.15,.15,.15,  // 61-70
  .15,.15,.15,.15,.15,.15,.15,.15,.15,.15,  // 71-80
  .15,.15,.15,.15,.15,.15,.15,.15,.15,.15,  // 81-90
  .15,.15,.15,.15,.15,.15,.15,.15,.15,.15   // 91-100
  };

  TLegend* leg;
 
  TString datafile="test/xsect.root";
  TString Name = "Total and elastic cross sections p + p in mb";
  TFile* f = new TFile(datafile);
  if(!f->IsOpen()) {
    cout <<  "File " << datafile  << " is opened" << endl;
    exit 1;
  }
  if(0 <= verbose) cout <<  "File " << datafile  << " is opened" << endl;

  TCanvas* c1 = new TCanvas("c1",Name,-1, 7, 900, 600);

  TH1F* hhh = gPad->DrawFrame(-1, 0.0001, 7, 10000. ,"p + p cross sections in mb");
  hhh->GetXaxis()->SetTitle("log10(E/MeV)");
  hhh->SetTitle(Name);
  hhh->Draw("SAME");
  gPad->SetLogy(1);
  gPad->Update();
  leg = new TLegend(.5, .2, 0.7, .4);
  leg->SetTextSize(.03);
  leg->SetHeader("Cross sections"); 

  TH1F* hh1 = (TH1F*)f->Get("0");
    hh1->SetLineColor(2);
    hh1->SetLineWidth(3);
    hh1->SetMarkerSize(.2);
    hh1->SetMarkerStyle(20);
    hh1->SetMarkerColor(2);
    leg->AddEntry(hh1, "Penelope", "pl");
    hh1->Draw("HIST SAME PL");

    TH1F* hh2 = (TH1F*)f->Get("h2");
    hh2->SetLineColor(3);
    hh2->SetLineWidth(3.);
    hh2->SetMarkerSize(.2);
    hh2->SetMarkerStyle(21);
    hh2->SetMarkerColor(3);
    leg->AddEntry(hh2, "Livermore", "pl");
    hh2->Draw("HIST SAME PL");
    
    TH1F* hh3 = (TH1F*)f->Get("h3");
    hh3->SetMarkerSize(.2);
    hh3->SetLineWidth(2);
    hh3->SetLineColor(4);
    hh3->SetMarkerStyle(22);
    hh3->SetMarkerColor(4);
    leg->AddEntry(hh3, "Bethe-Heitler-5D", "pl");
    leg->Draw("SAME");
    hh3->Draw("HIST SAME PL");

    TH1F* hh4 = (TH1F*)f->Get("h4");
    hh4->SetMarkerSize(.2);
    hh4->SetLineColor(6);
    hh4->SetLineWidth(2);
    hh4->SetMarkerStyle(22);
    hh4->SetMarkerColor(6);
    leg->AddEntry(hh4, "Bethe-Heitler", "pl");
    leg->Draw("SAME");
    hh4->Draw("HIST SAME PL");

    TH1F* hh5 = (TH1F*)f->Get("h5");
    hh5->SetMarkerSize(.2);
    hh5->SetLineWidth(1);
    hh5->SetLineColor(46);
    //    hh5->SetLineStyle(3);
    hh5->SetMarkerStyle(22);
    hh5->SetMarkerColor(46);
    leg->AddEntry(hh5, "Relativistic", "pl");
    leg->Draw("SAME");
    hh5->Draw("HIST SAME PL");
    /*
    TH1F* hh6 = (TH1F*)f.Get("h6");
    hh6->SetMarkerSize(.2);
    hh6->SetLineColor(7);
    hh6->SetLineWidth(1.);
    //    hh6->SetLineStyle(3);
    hh6->SetMarkerStyle(22);
    hh6->SetMarkerColor(7);
    leg->AddEntry(hh6, "Boldyshev", "pl");
    leg->Draw("SAME");
    hh6->Draw("HIST SAME PL");
    */

    cv->Update();
    name = "_" + s;
    cv->Print("plotXS/ApicXS" + name + ".png");
    //    gPad->Close(); 
    cv->Close();
  }
}


plotXS.py
#!/usr/bin/env python
#
# Plot XS pictures 
# A. Bagulya
# 23.03.2019
#

import os, sys
from ROOT import TROOT, gROOT, gStyle, gPad, TCanvas, TStyle, TPad
from ROOT import TFile, TH1, TLegend, TGraphAsymmErrors
from array import array
from math import log10

def plotXS():
    gROOT.Reset()
    gROOT.SetStyle('Plain')
    gROOT.SetBatch(True)

    HADSRC = readEnvVar('HADSRC')
    testpath = os.path.join(HADSRC,'HadrHN_class_test','test')
    plotpath = os.path.join(HADSRC,'HadrHN_class_test','plotXS')
    datapath = os.path.join(HADSRC,'HadrHN_class_test','data')

    if len(sys.argv) < 3:
        print '''Usage: python plotXS.py file.root 0/1'''
        sys.exit()
    else:
        fname = sys.argv[1]
        flag  = sys.argv[2]

    project = os.path.splitext(os.path.basename(fname))[0].split('_')[0]
    target = os.path.splitext(os.path.basename(fname))[0].split('_')[1]
    model_name = os.path.splitext(os.path.basename(fname))[0].split('_')[2]
    if model_name == 'kNS':
        model_name = 'NS'

    part = {'proton':'p', 'neutron':'n', 'pi+':'#pi^{+}', 'pi-':'#pi^{-}', 'kaon+':'K^{+}', 'kaon-':'K^{-}', 'gamma':'#g' , 'deuteron':'d', 'triton':'t', 'He3':'he3', 'alpha':'alpha'}

    global PLAB, SIG, errPLABl, errPLABh, errSIGl, errSIGh

    PLAB, SIG = array ( 'd' , [] ), array ( 'd' , [] )
    errPLABl, errPLABh, errSIGl, errSIGh = array ( 'd' , [] ), array ( 'd' , [] ), array ( 'd' , [] ), array ( 'd' , [] )

    c1 = TCanvas('c1', 'c1',200,100,900,600)
    gStyle.SetOptStat(0)
    c1.SetFillColor(0)
    c1.SetBorderMode(0)
    c1.SetBorderSize(0)
    c1.SetFrameBorderMode(0)

    hasDataEl = 1;
    hasDataIn = 1;
    if project == 'gamma':
        hasDataEl = 0;
    
    if project == 'alpha':
        hasDataEl = 0
        hasDataIn = 0;

    Name = 'Total, elastic and inelastic cross sections ' + part[project] + ' + ' + part[target] + ' in mb, model = ' + model_name
    if project == 'gamma':
        Name = 'Total cross sections gamma' + part[target] + ' in mb, model = ' + model_name

    if project == 'neutron':
        hhh = gPad.DrawFrame(1, 1., 7, 10000. ,'cross sections in mb')
    elif project == 'kaon+':
        hhh = gPad.DrawFrame(1, 1., 7, 100. ,'cross sections in mb')
    elif project == 'gamma':
        hhh = gPad.DrawFrame(1, 0.01, 7, 1. ,'cross sections in mb')
    elif project == 'proton' and flag == '1':
        hhh = gPad.DrawFrame(1, 1., 12, 1000. ,'cross sections in mb')
    else:
        hhh = gPad.DrawFrame(1, 1., 7, 1000. ,'cross sections in mb')

    hhh.SetTitle(Name)
    hhh.GetXaxis().SetTitle('log10(P/(MeV/c))')
    hhh.GetYaxis().SetTitle('#sigma, mb')
    hhh.GetXaxis().SetTitleOffset(1.2)
    hhh.GetYaxis().SetTitleOffset(1.2)
    hhh.Draw('AXIS SAME')

    gPad.SetLogy(1)
    gPad.Update()

    leg = TLegend(.77, .75, 0.97, .95)
    leg.SetTextSize(.03)
    leg.SetHeader('Cross sections') 

    rootflname = os.path.join(testpath, fname)
    print 'File to open ', rootflname
    f1 = TFile(rootflname)

    AddH1(f1, leg, 2, 'h0', 'Inelastic ' + model_name)
    c1.Update()
    if project != 'gamma':
        AddH1(f1, leg, 3, 'h1', 'Elastic ' + model_name)
        c1.Update()
        AddH1(f1, leg, 4, 'h2', 'Total ' + model_name)
        c1.Update()
#    AddH1(f1, leg, 5, 'h3', 'EM Elastic')
#    c1.Update()

    if project == 'pi+':
        datafile = os.path.join(datapath, 'rpp2016-pip' + part[target] + '_elastic.dat')
    elif project == 'pi-':
        datafile = os.path.join(datapath, 'rpp2016-pim' + part[target] + '_elastic.dat')
    elif project == 'kaon+':
        datafile = os.path.join(datapath, 'rpp2016-kp' + part[target] + '_elastic.dat')
    elif project == 'kaon-':
        datafile = os.path.join(datapath, 'rpp2016-km' + part[target] + '_elastic.dat')
    else:
        datafile = os.path.join(datapath, 'rpp2016-' + part[project] + part[target] + '_elastic.dat')

##    if project != 'gamma':
    if hasDataEl == 1:
        AddGraph(datafile, leg, 'data El PDG16', 1)

    if project == 'pi+':
        datafile = os.path.join(datapath, 'rpp2016-pip' + part[target] + '_total.dat')
    elif project == 'pi-':
        datafile = os.path.join(datapath, 'rpp2016-pim' + part[target] + '_total.dat')
    elif project == 'kaon+':
        datafile = os.path.join(datapath, 'rpp2016-kp' + part[target] + '_total.dat')
    elif project == 'kaon-':
        datafile = os.path.join(datapath, 'rpp2016-km' + part[target] + '_total.dat')
    elif project == 'gamma':
        datafile = os.path.join(datapath, 'rpp2016-gamma' + part[target] + '_total.dat')
    else:
        datafile = os.path.join(datapath, 'rpp2016-' + part[project] + part[target] + '_total.dat')

    if hasDataIn == 1:
        AddGraph(datafile, leg, 'data Tot PDG16', 2)

    leg.Draw('SAME')
    c1.Update() 

    if flag == '1':
        fout = os.path.join(plotpath, 'A_' + os.path.splitext(os.path.basename(fname))[0]) + '_cosmic.png' 
    else:
        fout = os.path.join(plotpath, 'A_' + os.path.splitext(os.path.basename(fname))[0]) + '.png'
    c1.Print(fout)
    c1.Close()

def readEnvVar(envar):
    if os.environ.get(envar) is None:
        print '''No environment variable %s''' % envar
        sys.exit()
    else:
        return os.environ.get(envar)

def openFile(fileName):
    global infile
    try:
        infile = open(fileName, 'r')
    except IOError:
        print 'Input file <',fileName, '> does not exist! Exit'
        sys.exit(2)

def AddH1(ff, leg, idx, hh, title):
    h1 = gROOT.FindObject(hh)
    h1.SetLineColor(idx)
    h1.SetLineWidth(2)
    h1.Draw('HISTO SAME C')
    leg.AddEntry(h1, title, 'l')

def readFile(ff):
    openFile(ff)
    lines = infile.readlines()
    num = int(lines[10])
    for line in lines[11:11+num]:
        plab = 1000.*float(line.split()[1])
        pmin = 1000.*float(line.split()[2])
        pmax = 1000.*float(line.split()[3])
        sig = float(line.split()[4])
        errsigh = float(line.split()[5])
        errsigl = float(line.split()[6])

        PLAB.append(log10(plab))
        errPLABl.append(0.)
        errPLABh.append(0.)
        SIG.append(sig)
        errSIGl.append(errsigh)
        errSIGh.append(errsigl)
    infile.close()

def AddGraph(ff, leg, title, idx):
     openFile(ff)
     readFile(ff)
     globals()['gr' + str(idx)] = TGraphAsymmErrors(len(PLAB),PLAB,SIG,errPLABl,errPLABh,errSIGl,errSIGh)
     globals()['gr' + str(idx)].SetMarkerStyle(19+idx)
     globals()['gr' + str(idx)].SetMarkerColor(1)
     globals()['gr' + str(idx)].SetMarkerSize(0.8)
     globals()['gr' + str(idx)].Draw('SAME P')
     leg.AddEntry(globals()['gr' + str(idx)],title,'p')

     del PLAB[:]
     del errPLABl[:]
     del errPLABh[:]
     del SIG[:]
     del errSIGl[:]
     del errSIGh[:]

###______________________________
if __name__ == '__main__':
    plotXS()


plotXSmodel_only.py
#!/usr/bin/env python
#
# Plot XS pictures 
# A. Bagulya
# 23.03.2019
#

import os, sys
from ROOT import TROOT, gROOT, gStyle, gPad, TCanvas, TStyle, TPad
from ROOT import TFile, TH1, TLegend, TGraphAsymmErrors
from array import array
from math import log10

def plotXS():
    gROOT.Reset()
    gROOT.SetStyle('Plain')
    gROOT.SetBatch(True)

    #HADSRC = readEnvVar('HADSRC')
    testpath = os.path.join('./','')
    plotpath = os.path.join('./','plotXS')
    datapath = os.path.join('./','data')

    if len(sys.argv) < 3:
        print('''Usage: python plotXS.py file.root 0/1''')
        sys.exit()
    else:
        fname = sys.argv[1]
        flag  = int(sys.argv[2])

    project = os.path.splitext(os.path.basename(fname))[0].split('_')[0]
    target = os.path.splitext(os.path.basename(fname))[0].split('_')[1]
    model_name = os.path.splitext(os.path.basename(fname))[0].split('_')[2]
    Z = os.path.splitext(os.path.basename(fname))[0].split('_')[3]
    A = os.path.splitext(os.path.basename(fname))[0].split('_')[4]
    if model_name == 'kNS':
        model_name = 'NS'
        
    if int(flag) == 1 and project != 'gamma':
        print('Flag 1 (Inelastic cross-section from 0 to 200 MeV) implemented only for gamma')
        exit()
        
    part = {'proton':'p', 'neutron':'n', 'pi+':'#pi^{+}', 'pi-':'#pi^{-}', 'kaon+':'K^{+}', 'kaon-':'K^{-}', 'gamma':'#g' , 'deuteron':'d', 'triton':'t', 'He3':'he3', 'alpha':'alpha'}

    global PLAB, SIG, errPLABl, errPLABh, errSIGl, errSIGh

    PLAB, SIG = array ( 'd' , [] ), array ( 'd' , [] )
    errPLABl, errPLABh, errSIGl, errSIGh = array ( 'd' , [] ), array ( 'd' , [] ), array ( 'd' , [] ), array ( 'd' , [] )

    c1 = TCanvas('c1', 'c1',200,100,900,600)
    gStyle.SetOptStat(0)
    c1.SetFillColor(0)
    c1.SetBorderMode(0)
    c1.SetBorderSize(0)
    c1.SetFrameBorderMode(0)

    hasDataEl = 0;
    hasDataIn = 0;
    if project == 'gamma':
        hasDataEl = 0;
    
    if project == 'alpha':
        hasDataEl = 0
        hasDataIn = 0;

    Name = 'Total, elastic and inelastic cross sections ' + project + ' + ' + target + ' in mb, model = ' + model_name
    if project == 'gamma':
        Name = 'Total cross sections gamma' + ' + '+target + A+ ' in mb, model = ' + model_name
        
    crossampl = [2,   1,   6,   10,  6,  20,  24,  26,  29,  30,  30,  30,  45,  45,  50,  60,  70,  80,  65,  90, 100,  100,  110,  110,  120,  120,  130,  130,  140,  140, 150,  150,  160,   170,   170,  200,   210,   220,  230,  240,   250, 260,   270,  280,  290, 300, 310, 320, 330, 340, 350,   360, 370,   380, 390, 400,   410,   420,   430,   440,  450,   460,   470, 480,   490,   500,   510,   520,   530, 540,  550,   560, 600, 600,   610,   620,   630, 650, 660,   680, 700, 720,  740,   760,   780,   800,   820,   840,   860, 880, 900]
    

    lbound = 1.;
    rbound = 100000;
    if project == 'neutron':
        hhh = gPad.DrawFrame(1, 1., 7, 10000. ,'cross sections in mb')
    elif project == 'kaon+':
        hhh = gPad.DrawFrame(1, 1., 7, 100. ,'cross sections in mb')
    elif project == 'gamma' and int(flag) == 0:
        hhh = gPad.DrawFrame(1., 1., 7., crossampl[int(Z)] ,'cross sections in mb')
    elif project == 'gamma' and int(flag) == 1:
        hhh = gPad.DrawFrame(lbound, 0.01, rbound, 4*crossampl[int(Z)] ,'cross sections in mb')
    elif project == 'proton':
        hhh = gPad.DrawFrame(1, 1., 12, 1000. ,'cross sections in mb')
    else:
        hhh = gPad.DrawFrame(1, 1., 7, 1000. ,'cross sections in mb')

    hhh.SetTitle(Name)
    if int(flag) == 0:
        hhh.GetXaxis().SetTitle('log10(P/(MeV/c))')
    else:
        hhh.GetXaxis().SetTitle('P, MeV/c')
    
    hhh.GetYaxis().SetTitle('#sigma, mb')
    hhh.GetXaxis().SetTitleOffset(1.2)
    hhh.GetYaxis().SetTitleOffset(1.2)
    hhh.Draw('AXIS SAME')
    if int(flag) == 0:
        gPad.SetLogy(1)
    else :
        gPad.SetLogx(1)
        gPad.SetLogy(1)
    gPad.Update()

    leg = TLegend(.77, .75, 0.97, .95)
    leg.SetTextSize(.03)
    leg.SetHeader('Cross sections') 

    rootflname = os.path.join(testpath, fname)
    print('File to open ', rootflname)
    f1 = TFile(rootflname)
    if int(flag) == 1:
        AddH1(f1, leg, 2, 'h4', 'Photonuclear Low Energy region ' + model_name)
    else:
        AddH1(f1, leg, 2, 'h0', 'Inelastic ' + model_name)
    c1.Update()
    if project != 'gamma':
        AddH1(f1, leg, 3, 'h1', 'Elastic ' + model_name)
        c1.Update()
        AddH1(f1, leg, 4, 'h2', 'Total ' + model_name)
        c1.Update()

    if int(flag) == 0:
       fout = os.path.join(plotpath, 'A_' + os.path.splitext(os.path.basename(fname))[0]) + '.png'
    else:
       fout = os.path.join(plotpath, 'A_' + os.path.splitext(os.path.basename(fname))[0])+'_'+str(lbound)+'_'+str(rbound) + 'MeV.png'
        
    c1.Print(fout)
    c1.Close()

def readEnvVar(envar):
    if os.environ.get(envar) is None:
        print('''No environment variable %s''' % envar)
        sys.exit()
    else:
        return os.environ.get(envar)

def openFile(fileName):
    global infile
    try:
        infile = open(fileName, 'r')
    except IOError:
        print('Input file <',fileName, '> does not exist! Exit')
        sys.exit(2)

def AddH1(ff, leg, idx, hh, title):
    h1 = gROOT.FindObject(hh)
    h1.SetLineColor(idx)
    h1.SetLineWidth(2)
    h1.Draw('HISTO SAME C')
    leg.AddEntry(h1, title, 'l')

def readFile(ff):
    openFile(ff)
    lines = infile.readlines()
    num = int(lines[10])
    for line in lines[11:11+num]:
        plab = 1000.*float(line.split()[1])
        pmin = 1000.*float(line.split()[2])
        pmax = 1000.*float(line.split()[3])
        sig = float(line.split()[4])
        errsigh = float(line.split()[5])
        errsigl = float(line.split()[6])

        PLAB.append(log10(plab))
        errPLABl.append(0.)
        errPLABh.append(0.)
        SIG.append(sig)
        errSIGl.append(errsigh)
        errSIGh.append(errsigl)
    infile.close()

def AddGraph(ff, leg, title, idx):
     openFile(ff)
     readFile(ff)
     globals()['gr' + str(idx)] = TGraphAsymmErrors(len(PLAB),PLAB,SIG,errPLABl,errPLABh,errSIGl,errSIGh)
     globals()['gr' + str(idx)].SetMarkerStyle(19+idx)
     globals()['gr' + str(idx)].SetMarkerColor(1)
     globals()['gr' + str(idx)].SetMarkerSize(0.4)
     globals()['gr' + str(idx)].Draw('SAME P')
     leg.AddEntry(globals()['gr' + str(idx)],title,'p')

     del PLAB[:]
     del errPLABl[:]
     del errPLABh[:]
     del SIG[:]
     del errSIGl[:]
     del errSIGh[:]

###______________________________
if __name__ == '__main__':
    plotXS()
